### ISSUE TITLE
Web Annotation Data Model

### ISSUE THREAD

**Issue by simonvarey** [Inclusio-Community/json-image-metadata#38]
Following a suggestion from @shepazu, we could use or adapt the Web Annotation Data Model for the `tags` field: https://www.w3.org/TR/annotation-model/

**Comment by shepazu** [comment]
I don't think the Web Annotation Data Model is suitable for `tags`. It's far too verbose, and doesn't offer any specific ontology for tags.

I do think the Web Annotation Data Model could be adopted/adopted for an as-yet-unspecified feature of JIM, the `annotations` block.

**Comment by simonvarey** [comment]
Thanks @shepazu for clearing that up. I had forgotten exactly why we were discussing the Web Annotation Data Model.

With this revision in mind, I'll change the issue to the following: Should we add an `annotations` block, and if so how much of the Web Annotation Data Model should we use for this feature?

**Comment by shepazu** [comment]
Yes, I think we should have an annotation block. I think annotations should be separate from the `dataset`, because a single annotation may refer to more than one datapoint, etc. So, we need to think about what the format of these annotations is. Do they only point to the JIM, or to the DOM, or both?

I'll go through the Web Annotation spec and see which keys we want to include. Since we're only doing text annotations, it will be a subset of their whole data model.

**Comment by shepazu** [comment]
Here are some of the JSON keys ("terms" in Web Annotation parlance) that seem directly relevant to our use of annotations.

```json
{
 "id": "http://example.org/anno5",
  "bodyValue": "Comment text",
  "target": "http://example.org/photo1"
  "creator": "http://example.org/user1",
  "created": "2015-01-28T12:00:00Z",
  "modified": "2015-01-29T09:00:00Z",
  "generator": "http://example.org/client1",
  "generated": "2015-02-04T12:00:00Z",
  "motivation": "commenting"
}
```


### Expanded entries

In many of these keys, the value can be a string (normally a URI/URL), or an object containing more details. 

For example, `creator` (a URI) can be expanded like so:

```json
  "creator": "http://example.org/user1"
``

```json
  "creator": {
    "id": "http://example.org/user1",
    "type": "Person",
    "name": "My Pseudonym",
    "nickname": "pseudo",
    "email_sha1": "58bad08927902ff9307b621c54716dcc5083e339"
  }
```

Similarly with `generator`:

```json
  "generator": "http://example.org/client1",
``

```json
  "generator": {
    "id": "http://example.org/client1",
    "type": "Software",
    "name": "Code v2.1",
    "homepage": "http://example.org/client1/homepage1"
  }
 ```

Confusingly (to me), the Working Group insisted that the base value **had to be** a URI, rather than a string, so they compromised by adding a `bodyValue` that was a shorthand for the `body` object:

```json
  "bodyValue": "Comment text",
```

Equivalent to:

 ```json
 "body": {
    "type": "TextualBody",
    "value": "Comment text",
    "format": "text/plain"
  },
```

This does add room for more keys like `language` (which is useful), `format` (which can be useful, though it should default to `text/plain`, which is the only case we're supporting), and `type` (which isn't very useful, especially if you don't support multimedia annotations, which we won't).

```json
   "body": {
    "type" : "TextualBody",
    "value" : "<p>j'adore !</p>",
    "format" : "text/html",
    "language" : "fr"
  }
 ```
 
### Selectors

There are several selectors available for the `target` key, none of which really capture what we want for JIM. Here's a couple of examples:

```json
"target": {
    "source": "http://example.org/page1.html",
    "selector": {
      "type": "CssSelector",
      "value": "#elemid > .elemclass + p"
    }
}
```
  
```json
  "target": {
    "source": "http://example.org/diskimg1",
    "selector": {
      "type": "DataPositionSelector",
      "start": 4096,
      "end": 4104
    }
  }
```

#### Extended selector

I'm thinking we could extend their selectors with a JIM-specific selector:

```json  
  "target": {
    "source": "http://example.org/chart1",
    "selector": {
      "type": "JIMSelector",
      "json": [
        "$.datasets[0].series[1].name",
        "$.datasets[0].series[1].records[23].*"
      ],
      "dom": "#datapoint-23"
    }
  }
}
```


### Constraints of Web Annotation syntax

Because Web Annotations was intended for "Semantic Web" usage, the model relies on [JSON-LD](https://en.wikipedia.org/wiki/JSON-LD), a JSON serialization of [RDF](https://en.wikipedia.org/wiki/Resource_Description_Framework).

This results in a data model that decouples values and adds in redirection links to resources, rather than directly stating values. IIRC, it sometimes requires nesting of resources in objects, rather than plain-text values, as with the case of the `body` object needing an "alias" of sorts with the `bodyText` object for the simplest (and most common) use case.

We don't want to perpetuate this unnecessary verbosity, especially since our annotations will all be text strings, so I suggest we simplify the syntax, while remaining compatible with the underlying data model. In other words, we should have a syntax that can be programmatically expanded into the full Web Annotation Data Model syntax.

Ironically, the overuse of linking in the Web Annotation Data Model did not result in one of the most pragmatic and useful use cases: having a single "reference block" containing the reusable parts of an annotation, for a document which has multiple annotations in it, leading to even more avoidable redundancy and verbosity.

For example, I think we could factor out the  `creator`, `generator`, and `motivation` objects (at least) into a referenceable reusable block.
